### DISCUSSION TITLE
Getserversideprops not working on Vercel with next-i18next

### DISCUSSION THREAD

**Discussion by hedysnike** [vercel/next.js#47187] [Help]
### Summary

it translates on localhost
![image](https://user-images.githubusercontent.com/117448065/225465516-0c4de815-e256-4a84-8345-0bf76241352c.png)

and on Vercel it only works if I use the getStaticProps
![image](https://user-images.githubusercontent.com/117448065/225465670-bd9c2561-4f68-4fd0-a5eb-a7a8ffbcf501.png)

I also require mobile agent so i can't really just use getstaticprops
This is my current code which works on localhost but not on vercel 
![image](https://user-images.githubusercontent.com/117448065/225465935-a2e2f232-99e9-4f11-8c14-837af39c641f.png)

**Comment by hedysnike** [comment]
this is code in case someone needs to copy


```
export const getServerSideProps: GetServerSideProps = async (context) => {
  const isMobile = context.req.headers["user-agent"]?.includes("Mobile");

  const locale = context.locale ?? 'ge';
  const translations = await serverSideTranslations(locale, [
    'common',
    'footer',
  ]);

  return {
    props: {
      isMobile,
      ...translations,
    },
  };
};

```



```
  export async function getStaticProps({ locale }: { locale: any}) {
    return {
      props: {
        ...(await serverSideTranslations(locale, [
          'common',
          'footer',
        ])),
      },
    }
  }
  

```

**Comment by hedysnike** [comment]
bump

**Comment by icyJoseph** [comment]
@hedysnika what happens when deployed to vercel? are the translations missing? does it crash?

**Reply by hedysnike** [reply]
Translations are missing instead translation names are shown 
for example: {t('registration')} will print "registration" instead of "Registration" or "რეგისტრაციის"
![image](https://user-images.githubusercontent.com/117448065/225674166-208625af-1cb3-4979-928e-9e7998bb4afc.png)
![image](https://user-images.githubusercontent.com/117448065/225674260-418deb32-71bc-4e2e-a303-952fa3ad443d.png)

No, it doesn't crash, doesn't show any errors only this error in console on live site:


react-i18next:: You will need to pass in an i18next instance by using initReactI18next
overrideMethod @ react_devtools_backend.js:2655
i @ _app-2075d3bc9c719811.js:8
s @ _app-2075d3bc9c719811.js:8
x @ _app-2075d3bc9c719811.js:8
v @ 966-0819c03af31c98fb.js:1
ab @ framework-281bd828f4ecdac7.js:9
i @ framework-281bd828f4ecdac7.js:9
oO @ framework-281bd828f4ecdac7.js:9
(anonymous) @ framework-281bd828f4ecdac7.js:9
oS @ framework-281bd828f4ecdac7.js:9
x @ framework-281bd828f4ecdac7.js:33
T @ framework-281bd828f4ecdac7.js:33

**Reply by hedysnike** [reply]
i added this to _app.tsx
```
import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';

i18n.use(initReactI18next).init({
    lng: 'ge',
    fallbackLng: 'ge',
    debug: true,
  });

```

now it gives me this error:
#i18next::backendConnector: No backend was added via i18next.use. Will not load resources.

**Reply by hedysnike** [reply]
![image](https://user-images.githubusercontent.com/117448065/225680754-0c8f3cf0-9379-444c-a2d6-d1eb00c7c6e8.png)

**Reply by icyJoseph** [reply]
Mmm do you think you could setup a small reproduction demo? I am not saying there's no issue, but I don't have so much time available to set it up and then help debugging. If you could make one then we can jump on it and try to figure out what's the problem. 

So far it'd seem that the translations are failing to start because of the serverless deploy done on Vercel, but I cannot find relevant info, only this one, https://github.com/i18next/next-i18next/issues/1917 and few others - I'd recommend asking over there too, this looks like a problem with that library + serverless deployment.

**Reply by hedysnike** [reply]
> Mmm do you think you could setup a small reproduction demo? I am not saying there's no issue, but I don't have so much time available to set it up and then help debugging. If you could make one then we can jump on it and try to figure out what's the problem.
> 
> So far it'd seem that the translations are failing to start because of the serverless deploy done on Vercel, but I cannot find relevant info, only this one, [i18next/next-i18next#1917](https://github.com/i18next/next-i18next/issues/1917) and few others - I'd recommend asking over there too, this looks like a problem with that library + serverless deployment.

Thank You Very Much I have already fixed the issue. I'll post below step by step for anyone who's still struggling and give recommendations.

It took me 2 days of depression followed by 20 hours of struggle and 4 Coffees as I'm still a I'm very new to programming in general

**Reply by icyJoseph** [reply]
Awesome! It's not the first time I see this issue though, would really love to see what fixed it.

Good job!

**Reply by hedysnike** [reply]
My setup is Next.js 13 / Typescript - run on vercel. (non experimental APP)
A lot from official example didn't work from me so I'll post the code from Scratch:

**Reply by hedysnike** [reply]
next-i18next.config.js

```
const path = require('path');

module.exports = {
    i18n: {
      locales: ['ge', 'en', 'ru'],
      defaultLocale: 'ge',
      localeDetection: false,
    },
    fallbackLng: { default: ['ge', 'en', 'ru' ]},
    defaultNs: 'shared',
    localePath: path.resolve('./public/locales'),
  };
  
```

next.config.mjs

```
const config = {
  reactStrictMode: true,
  i18n: {
    locales: ['ge', 'en', 'ru'],
    defaultLocale: 'ge',
    localeDetection: false,
  },
};

export default config;

```

_app

```
import { appWithTranslation } from "next-i18next";

export default appWithTranslation(MyApp);
```

What I did at first before getting "getServerSideProps" to work was basically, I removed "Mobile Agent" all together and rewrote CSS, it made the Website faster so I decided to keep it that way.

below is code for Typescript "getStaticProps".
```
export async function getStaticProps({ locale }: { locale: any}) {
  return {
    props: {
      ...(await serverSideTranslations(locale, [
        'common',
        'footer',
      ])),
    },
  }
}

```

I had issue of getting "getStaticProps" to work on dynamic page, But luckily this code worked for me:

```
export const getServerSideProps = async (context: GetServerSidePropsContext) => {
  const { locale } = context;
  
  if (!locale) {
    return {
      notFound: true,
    };
  }
  
  const response = await fetch(`${process.env.NEXT_PUBLIC_BASE_URL}/api/YOUR API NAME`);
  const user = await response.json();

  return {
    props: {
      user,
      ...(await serverSideTranslations(locale, ['home', 'common']))
    }
  };
};
```
IMPORTANT: For above code to work on dynamic page you also have to add Environment Variables on Vercel:
key="NEXT_PUBLIC_BASE_URL"
Value="Your Website Url" 
Value example="https://google.com"

On localhost You can just add same thing to .env File: example:
NEXT_PUBLIC_BASE_URL="https://google.com"

It will actually work with adding live website in URL on localhost as resources are available for fetching.

For non Dynamic Pages, I assume code would look like this

```
export const getServerSideProps = async (context: GetServerSidePropsContext) => {
  const { locale } = context;
  
  if (!locale) {
    return {
      notFound: true,
    };
  }
  

  return {
    props: {
      ...(await serverSideTranslations(locale, ['home', 'common']))
    }
  };
};

```

**Reply by hedysnike** [reply]
For anyone Following literally from scratch you also have to create locales folder in public folder.

In locales folder you can create folder for each language and inside that language folder "common.json"
on individual pages you just import translation afterwards and multilingual site just works.

```

export default function (){
const { t } = useTranslation('common')

return(
<div>{t('home')} </div>
)
}
```
